src/lib.rs

```diff
@@ -8,6 +8,197 @@ pub mod qdrant;
+    extract!(DoubleValue, is_double, as_double, f64);
+    extract!(StringValue, is_str, as_str, ref String);
+    extract!(ListValue, is_list, as_list, ref ListValue);
+    extract!(StructValue, is_struct, as_struct, ref Struct);
+
+    #[cfg(feature = "serde")]
+    /// convert this into a `serde_json::Value`
+    pub fn into_json(self) -> serde_json::Value {
+        use serde_json::Value as JsonValue;
+        match self.kind {
+            Some(BoolValue(b)) => JsonValue::Bool(b),
+            Some(IntegerValue(i)) => JsonValue::from(i),
+            Some(DoubleValue(d)) => JsonValue::from(d),
+            Some(StringValue(s)) => JsonValue::String(s),
+            Some(ListValue(vs)) => vs.values.into_iter().map(Value::into_json).collect(),
+            Some(StructValue(s)) => s
+                .fields
+                .into_iter()
+                .map(|(k, v)| (k, v.into_json()))
+                .collect(),
+            Some(NullValue(_)) | None => JsonValue::Null,
+        }
+    }
+}
+
+#[cfg(feature = "serde")]
+impl From<Value> for serde_json::Value {
+    fn from(value: Value) -> Self {
+        value.into_json()
+    }
+}
+
+impl Display for Value {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        match &self.kind {
+            Some(BoolValue(b)) => write!(f, "{}", b),
+            Some(IntegerValue(i)) => write!(f, "{}", i),
+            Some(DoubleValue(v)) => write!(f, "{}", v),
+            Some(StringValue(s)) => write!(f, "{:?}", s),
+            Some(ListValue(vs)) => {
+                let mut i = vs.values.iter();
+                write!(f, "[")?;
+                if let Some(first) = i.next() {
+                    write!(f, "{}", first)?;
+                    for v in i {
+                        write!(f, ",{}", v)?;
+                    }
+                }
+                write!(f, "]")
+            }
+            Some(StructValue(s)) => {
+                let mut i = s.fields.iter();
+                write!(f, "{{")?;
+                if let Some((key, value)) = i.next() {
+                    write!(f, "{:?}:{}", key, value)?;
+                    for (key, value) in i {
+                        write!(f, ",{:?}:{}", key, value)?;
+                    }
+                }
+                write!(f, "}}")
+            }
+            _ => write!(f, "null"),
+        }
+    }
+}
+
+/// An error for failed conversions (e.g. calling `String::try_from(v)`
+/// on an integer value)
+pub struct NotA<T> {
+    marker: PhantomData<T>,
+}
+
+macro_rules! not_a {
+    ($ty:ty) => {
+        impl Error for NotA<$ty> {}
+
+        impl Debug for NotA<$ty> {
+            fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+                write!(f, "{}", self)
+            }
+        }
+
+        impl Display for NotA<$ty> {
+            fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+                f.write_str(concat!("not a ", stringify!($ty)))
+            }
+        }
+    };
+}
```

Comment:
That won't work because we cannot `stringify!($ty)` outside a macro. But I also think that being specific with the types makes for a clearer interface here.